File: TrackList/Managers /TagLibReader/TLTagLibFile.m
//
//  TLTagLibFile.m
//  TrackList
//
//  Реализация обёртки для чтения тегов через TagLib (C API)
//  Используется в Swift через Bridging Header
//
//  Created by Pavel Fomin on 21.06.2025.
//

#import <Foundation/Foundation.h>
#import "TLTagLibFile.h"
#import <tag_c.h>


static inline void TLTagLibSetup(void) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        taglib_set_string_management_enabled(false);
    });
}


// MARK: - Реализация класса результата

@implementation TLTagLibFileResult
// Пустая реализация — все свойства определены в заголовочном файле

@end


// MARK: - Главная функция чтения тегов

// Читает метаданные из указанного аудиофайла с помощью TagLib
// Возвращает объект TLTagLibFileResult или nil при ошибке
TLTagLibFileResult *_readMetadata(NSString *filePath) {
    
    TLTagLibSetup();
    
    const char *cPath = [filePath fileSystemRepresentation];
    TagLib_File *file = taglib_file_new(cPath);
    if (!file) return nil;

    TagLib_Tag *tag = taglib_file_tag(file);
    if (!tag) {
        taglib_file_free(file);
        return nil;
    }

    TLTagLibFileResult *result = [TLTagLibFileResult new];

    char *title = taglib_tag_title(tag);
    char *artist = taglib_tag_artist(tag);
    char *album = taglib_tag_album(tag);
    char *genre = taglib_tag_genre(tag);
    char *comment = taglib_tag_comment(tag);

    if (title && strlen(title) > 0) result.title = [NSString stringWithUTF8String:title];
    if (artist && strlen(artist) > 0) result.artist = [NSString stringWithUTF8String:artist];
    if (album && strlen(album) > 0) result.album = [NSString stringWithUTF8String:album];
    if (genre && strlen(genre) > 0) result.genre = [NSString stringWithUTF8String:genre];
    if (comment && strlen(comment) > 0) result.comment = [NSString stringWithUTF8String:comment];

    // Обрабатываем обложку ДО освобождения файла
    TagLib_Complex_Property_Attribute ***props = taglib_complex_property_get(file, "PICTURE");
    if (props) {
        TagLib_Complex_Property_Picture_Data picture = {0};
        taglib_picture_from_complex_property(props, &picture);

        if (picture.data && picture.size > 0) {
            void *copy = malloc(picture.size);
            if (copy) {
                memcpy(copy, picture.data, picture.size);
                NSData *imageData = [NSData dataWithBytesNoCopy:copy length:picture.size freeWhenDone:YES];
                result.artworkData = imageData;
            }
        }
        taglib_complex_property_free(props);
    }

    // Очистка
    // taglib_tag_free_strings(); // больше не вызываем
    taglib_file_free(file);

    return result;
}

File: TrackList/Managers /TagLibReader/TLTagLibFile.swift
//
//  TLTagLibFile.swift
//  TrackList
//
//  Низкоуровневая Swift-обёртка над C API TagLib.
//  Предоставляет унифицированный доступ к аудиотегам файла
//  без привязки к конкретному сценарию использования.
//
//  Используется различными слоями приложения:
//  - RuntimeMetadataParser (минимальный runtime-набор тегов)
//  - TrackTagInspector (расширенный набор тегов для инспекции)
//
//  Не содержит UI-логики и не знает о потребителях данных.
//  Отдаёт «сырые» значения тегов в нейтральном формате.
//
//  Created by Pavel Fomin on 21.06.2025.
//

import Foundation

/// Обёртка над C API TagLib для чтения тегов аудиофайла.
/// Является общим low-level адаптером для всех read-сценариев.
public class TLTagLibFile {
    
    /// Путь к аудиофайлу на диске
    private let filePath: String
    
    /// Инициализация с использованием URL аудиофайла
    public init(fileURL: URL) {
        self.filePath = fileURL.path
    }
    
    // MARK: - Внутреннее представление тегов
    
    /// Нейтральная структура с тегами, считанными напрямую из TagLib.
    /// Не является runtime-моделью и не предназначена для прямого использования в UI.
    public struct ParsedMetadata {
        public let title: String?
        public let artist: String?
        public let album: String?
        public let genre: String?
        public let comment: String?
        public let artworkData: Data?
    }
    
    // MARK: - Чтение тегов из TagLib
    
    /// Считывает доступные теги из файла через TagLib.
    /// Возвращает «сырые» данные без форматирования и бизнес-логики.
    public func readMetadata() -> ParsedMetadata? {
        
        // Вызов C-функции из Objective-C обёртки (TLTagLibFile.m)
        guard let result = _readMetadata(filePath) else {
            return nil
        }

        // Преобразование в Swift-структуру
        return ParsedMetadata(
            title: result.title,
            artist: result.artist,
            album: result.album,
            genre: result.genre,
            comment: result.comment,
            artworkData: result.artworkData
        )
    }
    
    // MARK: - Runtime-адаптация
    
    /// Преобразует считанные теги в runtime-модель TrackMetadata.
    /// Используется runtime-парсером для списка и плеера.
    /// Не применяется для инспекции или редактирования тегов.
    func readMetadata(duration: TimeInterval?) -> TrackMetadata {
        
        // Если TagLib не смог прочитать теги — возвращаем пустую runtime-модель
        guard let parsed = readMetadata() else {
            return TrackMetadata(
                artist: nil,
                title: nil,
                album: nil,
                artworkData: nil,
                duration: duration,
                isCustomFormat: true
            )
        }

        // Формирование runtime-модели с учётом длительности
        return TrackMetadata(
            artist: parsed.artist,
            title: parsed.title,
            album: parsed.album,
            artworkData: parsed.artworkData,
            duration: duration,
            isCustomFormat: true
        )
    }
}
